pythonlib/src/mill/pythonlib/RuffModule.scala

```diff
@@ -0,0 +1,80 @@
+package mill.pythonlib
+
+import mill._
+
+/**
+ * Linting and formatting functionality provided by [ruff](https://docs.astral.sh/ruff/).
+ */
+trait RuffModule extends PythonModule {
+
+  override def pythonToolDeps = Task {
+    super.pythonToolDeps() ++ Seq("ruff>=0.9.3")
+  }
+
+  /**
+   * Configuration file to use when running ruff. If this file does not exist,
+   * ruff will use the default settings.
+   */
+  def ruffConfigFile: T[PathRef] = Task.Source(millSourcePath / "ruff.toml")
+
+  /**
+   * Global command line options to pass to ruff. These are passed in before any
+   * command-supplied arguments.
+   */
+  def ruffOptions: T[Seq[String]] = Task { Seq.empty[String] }
+
+  // this is an anonymous task, since we want to return a PathRef as
+  // os.Shellable, instead of only the path
+  private def configArgs: Task[Seq[os.Shellable]] = Task.Anon {
+    val cfg = ruffConfigFile()
+    if (os.exists(cfg.path)) Seq[os.Shellable]("--config", cfg) else Seq.empty[os.Shellable]
+  }
```

Comment:
`os.Shellable` is basically an implicit constructor for conveniently passing `Seq[String]`s, so we can probably just use `T[Seq[String]]` here instead

Reply:
The reason why I didn't go with a plain string seq, was to prevent caching: essentially, I want factor out the argument *string* `--config, <path>`, but also want to invalidate the result if the *contents* of the config file change. I didn't think this through indeed and somehow thought that shellable will handle this (but of course that's wrong).

Reply:
You can have helper functions that return anonymous tasks, or anoynymous tasks which return a function type

If you want to make a task that invalidates whenever the file changes, you want a Task.Input

Reply:
> You can have helper functions that return anonymous tasks, or anoynymous tasks which return a function type

This sounds like what I'm looking for, thanks!

> If you want to make a task that invalidates whenever the file changes, you want a Task.Input

But this wouldn't work if I want the dependents of the task to be invalidated, right? 

```
def foo = Task.Input { "a" }
def bar = Task { foo() }
```

If I run `bar()`, then change "a" to "b", then re-run `bar()`, it would not actually be executed. Only `foo` would be reevaluated since it's an input, but since its output hasn't changed, bar would not be rerun afterwards.

Reply:
If you use a Task.Source pointing at the config file, then the contents of the config file changing will cause the returned PathRef to have a different hash, and any task _immediately_ downstream of the config file will be forced to re-evaluate. 

_Transitive_ downstream tasks then may or may not invalidate depending on whether the return value of the intermediate tasks changed, which is probably what you want most of the time

Reply:
Right, what was confusing me was that if I factored out the the command line construction (which contains only the file path) as an intermediary to the actual file contents, then it would effectively absorb any changes in the cache chain.

However, after some more reading through the task source code and testing, I think that a Task.Anon is actually sufficient here; there's no need for fancy function-returning tasks.

What I missed was that the results of Anon tasks are never cached, so if an upstream change causes an Anon task to be re-run, its dependents will also be re-run.

Thanks for your explanations!